apps/local_ledger_db/lib/local_ledger_db/entry.ex

```diff
@@ -1,133 +1,166 @@
+      raise InsufficientFundsError,
+        message: InsufficientFundsError.error_message(current_amount, attrs)
+    end
+
+    :ok
   end
 
   @doc """
-  Get a entry using one or more fields.
+  Calculate the total wallets for all the specified tokens associated
+  with the given address.
   """
-  @spec get_by(keyword() | map(), keyword()) :: %Entry{} | nil
-  def get_by(map, opts \\ []) do
-    query = Entry |> Repo.get_by(map)
+  def calculate_all_balances(address, options \\ %{}) do
+    options = Map.put_new(options, :token_id, :all)
+    credits = sum(address, Entry.credit_type(), options)
+    debits = sum(address, Entry.debit_type(), options)
 
-    case opts[:preload] do
-      nil -> query
-      preload -> Repo.preload(query, preload)
-    end
+    credits |> subtract(debits) |> format()
   end
 
-  @doc """
-  Helper function to get a entry with an idempotency token and loads all the required
-  associations.
-  """
-  @spec get_by_idempotency_token(String.t()) :: %Entry{} | nil
-  def get_by_idempotency_token(idempotency_token) do
-    get_by(
-      %{
-        idempotency_token: idempotency_token
-      },
-      preload: [:transactions]
-    )
+  defp subtract(credits, debits) do
+    tokens = Enum.uniq(Map.keys(credits) ++ Map.keys(debits))
+
+    Enum.map(tokens, fn token ->
+      {token, (credits[token] || 0) - (debits[token] || 0)}
+    end)
   end
 
-  def get_or_insert(%{idempotency_token: idempotency_token} = attrs) do
-    case get_by_idempotency_token(idempotency_token) do
-      nil ->
-        insert(attrs)
+  defp format(wallets), do: Enum.into(wallets, %{})
 
-      entry ->
-        {:ok, entry}
-    end
+  defp sum(address, type, options) do
+    address
+    |> build_sum_query(type, options)
+    |> Repo.all()
+    |> Enum.into(%{}, fn {k, v} -> {k, Decimal.to_integer(v)} end)
   end
 
-  @doc """
-  Insert an entry and its transactions.
-  """
-  def insert(attrs) do
-    opts = [on_conflict: :nothing, conflict_target: :idempotency_token]
+  defp build_sum_query(address, type, %{token_id: id, since: since, upto: upto}) do
+    address
+    |> build_sum_query(type, %{token_id: id})
+    |> where([e], e.inserted_at > ^since)
+    |> where([e], e.inserted_at <= ^upto)
+  end
+
+  defp build_sum_query(address, type, %{token_id: id, since: since}) do
+    address
+    |> build_sum_query(type, %{token_id: id})
+    |> where([e], e.inserted_at > ^since)
+  end
 
-    %Entry{}
-    |> changeset(attrs)
-    |> do_insert(opts)
+  defp build_sum_query(address, type, %{token_id: id, upto: upto}) do
+    address
+    |> build_sum_query(type, %{token_id: id})
+    |> where([e], e.inserted_at <= ^upto)
   end
 
-  defp do_insert(changeset, opts) do
-    case Repo.insert(changeset, opts) do
-      {:ok, entry} ->
-        entry.idempotency_token
-        |> get_by_idempotency_token()
-        |> handle_retrieval_result()
+  defp build_sum_query(address, type, %{token_id: :all}) do
```

Comment:
I was like "where is this all coming from?", then I remember we swapped them :D

Reply:
Yup. No code changes to these files. Only renames.